File: android-kotlin/app/src/main/java/io/github/steeb_k/commune/Updates.kt
// Keeping this installation current.
//
// The core does the same work here as on the desktop — read the signed
// release manifest, decide whether what it offers is newer, fetch the
// artifact and check its digest — and this is the part Android owns: asking
// at a sensible moment, and handing the APK to the system package installer.
//
// Installing is deliberately not something this code does. Android will only
// replace an installed application with one signed by the same key, and it is
// the system that enforces that, not us. The updater's whole security model
// rests on it, which is why the release keystore in
// ../../../../../../keystore.properties is the one irreplaceable file in the
// project: an APK signed by anything else cannot upgrade what is installed.
package io.github.steeb_k.commune

import android.app.Activity
import android.content.Context
import android.content.Intent
import android.net.Uri
import android.provider.Settings
import android.util.Log
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import androidx.core.content.FileProvider
import io.github.steeb_k.commune.core.CoreException
import io.github.steeb_k.commune.core.FfiRelease
import io.github.steeb_k.commune.core.FfiUpdateSettings
import io.github.steeb_k.commune.core.UpdateProgressListener
import io.github.steeb_k.commune.core.appBuildNumber
import io.github.steeb_k.commune.core.appVersion
import io.github.steeb_k.commune.core.checkForUpdate
import io.github.steeb_k.commune.core.downloadUpdate
import io.github.steeb_k.commune.core.setUpdateChannel
import io.github.steeb_k.commune.core.setUpdateCheckAutomatically
import io.github.steeb_k.commune.core.setUpdateSkippedVersion
import io.github.steeb_k.commune.core.updateSettings
import java.io.File
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext

private const val TAG = "CommuneUpdates"

/// What the updater is doing, and what it last found.
enum class UpdateState {
    /// Nothing has been asked yet this run.
    Idle,

    /// A check is in flight.
    Checking,

    /// The feed was read and this build is current.
    UpToDate,

    /// There is a newer release to install.
    Available,

    /// The APK is being fetched.
    Downloading,

    /// The package installer has been handed the APK.
    Installing,

    /// The last attempt did not finish.
    Failed,
}

/// The state of this installation with respect to the release feed.
///
/// One of these, like the desktop's `Updates` object, because there is one
/// installation: the settings rows observe it rather than each doing their
/// own check.
object Updates {
    /// What the updater is doing.
    var state by mutableStateOf(UpdateState.Idle)
        private set

    /// A sentence describing [state], for a row's subtitle.
    var status by mutableStateOf("")
        private set

    /// How much of the download has arrived, from 0 to 1.
    var progress by mutableStateOf(0f)
        private set

    /// The version of the release that is available, when one is.
    var availableVersion by mutableStateOf<String?>(null)
        private set

    /// Where a person can read what the available release changed.
    var releaseNotesUrl by mutableStateOf<String?>(null)
        private set

    /// The settings as they were last read.
    var settings by mutableStateOf<FfiUpdateSettings?>(null)
        private set

    /// The version of a release an automatic check found and has not yet
    /// been shown for outside Settings — a Snackbar's cue, cleared by
    /// [dismissAnnouncement] once it has been shown. `null` most of the
    /// time: only a check nobody asked for, that found a release nobody
    /// skipped, sets this.
    var announcement by mutableStateOf<String?>(null)
        private set

    /// The version this build reports as its own.
    val version: String by lazy { appVersion() }

    /// The build number this build reports as its own — the number the
    /// updater orders two builds of one version by.
    val buildNumber: ULong by lazy { appBuildNumber() }

    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.Main)
    private var release: FfiRelease? = null

    /// Whether [start] has already run once.
    ///
    /// `start()` is called from the main activity as soon as the core is up
    /// so that an installation nobody opens Settings on still checks, and
    /// `UpdateRows()` also calls it (so Settings works even if that call
    /// somehow never ran) — this makes the second call a no-op instead of a
    /// second check racing the first.
    private var started = false

    /// Read the settings, and check if one is due.
    ///
    /// Called once the core is up. An automatic check that cannot reach the
    /// feed says nothing: a phone with no signal is not a problem the user
    /// needs to be told about. Safe to call more than once; only the first
    /// call does anything.
    fun start() {
        if (started) return
        started = true

        scope.launch {
            val current = withContext(Dispatchers.IO) { updateSettings() }
            settings = current

            if (current.checkAutomatically && current.checkDue) {
                check(userInitiated = false)
            }
        }
    }

    /// Ask the feed what the current release is.
    fun check(userInitiated: Boolean) {
        if (state == UpdateState.Checking ||
            state == UpdateState.Downloading ||
            state == UpdateState.Installing
        ) {
            return
        }

        state = UpdateState.Checking
        status = "Checking…"

        scope.launch {
            try {
                val result = checkForUpdate()
                val current = withContext(Dispatchers.IO) { updateSettings() }
                settings = current

                val available = result.available
                // A release the user asked not to hear about again is not
                // announced by a check nobody asked for — but a press of
                // "Check Now" always gets an answer, skipped or not, which
                // is what makes the skip reversible rather than permanent.
                val isSkipped = !userInitiated &&
                    available != null &&
                    current.skippedVersion == available.version

                if (available?.asset == null || isSkipped) {
                    release = null
                    availableVersion = null
                    releaseNotesUrl = null
                    state = UpdateState.UpToDate
                    status = "Commune is up to date."
                } else {
                    release = available
                    availableVersion = available.version
                    releaseNotesUrl = available.notesUrl
                    state = UpdateState.Available
                    status = "Commune ${available.version} is available."

                    if (!userInitiated) {
                        announcement = available.version
                    }
                }
            } catch (error: Exception) {
                Log.w(TAG, "Could not check for updates", error)

                if (userInitiated) {
                    state = UpdateState.Failed
                    // `error.message` is not this: a `CoreException` other
                    // than `Failed` carries whatever `.message` a bindings
                    // internal-error class defaults to, and `Failed.message`
                    // itself is a debug rendering ("msg=...") rather than the
                    // core's user-facing sentence, which lives in `.msg`.
                    status = (error as? CoreException.Failed)?.msg
                        ?: "Could not check for updates."
                } else {
                    state = UpdateState.Idle
                    status = ""
                }
            }
        }
    }

    /// Stop offering the release the last check found, and put the row back
    /// to up to date — the same thing a check that found this version
    /// skipped would show.
    fun skip() {
        val version = release?.version ?: return

        scope.launch {
            withContext(Dispatchers.IO) { setUpdateSkippedVersion(version) }
            settings = withContext(Dispatchers.IO) { updateSettings() }
            release = null
            availableVersion = null
            releaseNotesUrl = null
            state = UpdateState.UpToDate
            status = "Commune is up to date."
        }
    }

    /// The last automatic-check announcement has been shown; do not show it
    /// again.
    fun dismissAnnouncement() {
        announcement = null
    }

    /// Turn automatic checks on or off.
    fun setCheckAutomatically(enabled: Boolean) {
        scope.launch {
            withContext(Dispatchers.IO) { setUpdateCheckAutomatically(enabled) }
            settings = withContext(Dispatchers.IO) { updateSettings() }
        }
    }

    /// Follow the named channel, and look at it straight away.
    fun setChannel(channel: String) {
        scope.launch {
            withContext(Dispatchers.IO) { setUpdateChannel(channel) }
            settings = withContext(Dispatchers.IO) { updateSettings() }
            check(userInitiated = true)
        }
    }

    /// Download the available release and offer it to the package installer.
    ///
    /// The user has to have allowed this application to install packages.
    /// There is no way to grant that from here, so the first time through
    /// this opens the settings page that can and stops; the next press gets
    /// as far as the installer.
    fun install(activity: Activity) {
        val asset = release?.asset ?: return

        if (state == UpdateState.Downloading || state == UpdateState.Installing) {
            return
        }

        if (!activity.packageManager.canRequestPackageInstalls()) {
            status = "Allow Commune to install apps, then press Update again."
            activity.startActivity(
                Intent(
                    Settings.ACTION_MANAGE_UNKNOWN_APP_SOURCES,
                    Uri.parse("package:${activity.packageName}"),
                ),
            )
            return
        }

        progress = 0f
        state = UpdateState.Downloading
        status = "Downloading…"

        scope.launch {
            try {
                val directory = File(activity.cacheDir, "updates")
                val listener = object : UpdateProgressListener {
                    override fun progress(downloaded: ULong, total: ULong) {
                        val fraction =
                            if (total == 0UL) 0f else downloaded.toFloat() / total.toFloat()

                        // Called from the download's own thread, often; the
                        // state it writes belongs to the main one.
                        scope.launch {
                            progress = fraction
                            status = "Downloading… ${(fraction * 100).toInt()}%"
                        }
                    }
                }

                val path = downloadUpdate(asset, directory.absolutePath, listener)

                state = UpdateState.Installing
                status = "Opening the installer…"
                offerToInstaller(activity, File(path))
            } catch (error: Exception) {
                Log.w(TAG, "Could not download the update", error)
                state = UpdateState.Failed
                status = (error as? CoreException.Failed)?.msg
                    ?: "Could not download the update."
            }
        }
    }

    /// The activity that handed an APK to the package installer has been
    /// resumed.
    ///
    /// `install()` starts the installer with `startActivity` and never
    /// learns what happened to it: the sheet can be cancelled, or opened and
    /// closed without a dialog, and neither tells this object anything.
    /// Before this existed, the row was then stuck on "Opening the
    /// installer…" with both buttons dead (`check()` returns early while
    /// `state == Installing`) until the process was killed and restarted. An
    /// app that is resumed while it still thinks it is being replaced was not
    /// replaced, so the release stays available — it is still held in
    /// [release] — and the row goes back to announcing it rather than
    /// staying stuck.
    fun installerReturned() {
        if (state != UpdateState.Installing) return

        val version = release?.version ?: run {
            state = UpdateState.Idle
            status = ""
            return
        }

        state = UpdateState.Available
        status = "Commune $version is available."
    }

    /// Hand the downloaded APK to the system package installer.
    private fun offerToInstaller(context: Context, apk: File) {
        val uri: Uri =
            FileProvider.getUriForFile(context, "${context.packageName}.updates", apk)

        context.startActivity(
            Intent(Intent.ACTION_VIEW).apply {
                setDataAndType(uri, "application/vnd.android.package-archive")
                addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION)
                addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
            },
        )
    }
}

File: src/updates.rs
//! Keeping this installation current.
//!
//! The part that is the same on every platform — reading the release feed,
//! checking its signature, deciding whether what it offers is newer — is
//! [`commune_core::updates`]. This is the application's half: when to ask,
//! what to say about the answer, and how to hand the downloaded artifact to
//! the platform that knows how to install it.
//!
//! There is one of these, owned by [`crate::Application`], because there is
//! one installation. The settings row and the toast are two views of it
//! rather than two things that each check.
//!
//! # When it stays quiet
//!
//! An update that cannot be installed must not be announced, so the check is
//! not made at all unless this build is one that could replace itself:
//!
//! * Inside Flatpak, and on Linux generally, the app store owns this.
//! * On Windows, unless the executable is under `%LOCALAPPDATA%\Programs` —
//!   anywhere else is a build directory or an unpacked `.zip`, and running an
//!   installer would add a second copy rather than replace this one.
//! * On macOS, unless the bundle carries a real Team ID. An ad-hoc signature
//!   means a developer's own build, and replacing it with a release would
//!   change the identity under every Keychain item it has stored.

use std::{cell::Cell, path::Path, sync::Arc, time::Duration};

use adw::{prelude::*, subclass::prelude::*};
use commune_core::{
    UserFacingError,
    updates::{self, Channel, DownloadProgress, Release, UpdateSettings},
};
use futures_channel::mpsc;
use futures_util::StreamExt;
use gettextrs::gettext;
use gtk::{glib, glib::clone};
use tracing::{debug, error, warn};

use crate::{gettext_f, spawn, spawn_tokio, utils::DataType};

/// How long after startup the first automatic check waits.
///
/// Long enough to be past the first sync, so that a check never competes with
/// the thing the user actually opened the application for.
const FIRST_CHECK_DELAY: Duration = Duration::from_secs(90);

/// How often the timer comes back to ask whether a check is due.
///
/// Not the interval between checks — [`UpdateSettings::is_check_due()`] owns
/// that, and it has to, because it is the only thing that survives a restart.
/// This is just how often a long-running process wakes up to look.
const TIMER_INTERVAL: Duration = Duration::from_hours(1);

/// Where an update is being downloaded to.
fn download_dir() -> std::path::PathBuf {
    DataType::Cache.dir_path().join("updates")
}

/// What the updater is doing, and what it last found.
#[derive(Debug, Default, Clone, Copy, PartialEq, Eq, glib::Enum)]
#[enum_type(name = "UpdateState")]
pub enum UpdateState {
    /// Nothing has been asked yet this run.
    #[default]
    Idle,
    /// A check is in flight.
    Checking,
    /// The feed was read and this build is current.
    UpToDate,
    /// There is a newer release to install.
    Available,
    /// The artifact is being fetched.
    Downloading,
    /// The artifact is on disk and the installer is starting.
    Installing,
    /// The last attempt did not finish.
    Failed,
}

mod imp {
    use super::*;

    #[derive(Debug, Default, glib::Properties)]
    #[properties(wrapper_type = super::Updates)]
    pub struct Updates {
        /// What the updater is doing.
        #[property(get, builder(UpdateState::default()))]
        pub state: Cell<UpdateState>,
        /// The version of the release that is available, when one is.
        #[property(get)]
        pub available_version: RefCell<String>,
        /// How much of the download has arrived, from 0 to 1.
        #[property(get)]
        pub progress: Cell<f64>,
        /// A sentence describing [`Self::state`], for a row's subtitle.
        #[property(get)]
        pub status: RefCell<String>,
        /// Whether this installation could install an update at all.
        #[property(get)]
        pub supported: Cell<bool>,
        /// The release the last check found.
        pub release: RefCell<Option<Release>>,
        /// The stored settings, read once.
        ///
        /// Kept here rather than loaded per call: every read of them is a
        /// `GSettings` lookup and a JSON parse, and this object is the only
        /// thing in the process that writes them.
        pub settings: RefCell<UpdateSettings>,
    }

    use std::cell::RefCell;

    #[glib::object_subclass]
    impl ObjectSubclass for Updates {
        const NAME: &'static str = "Updates";
        type Type = super::Updates;
    }

    #[glib::derived_properties]
    impl ObjectImpl for Updates {
        fn constructed(&self) {
            self.parent_constructed();

            self.settings.replace(UpdateSettings::load());
            self.supported.set(super::is_supported());
            self.obj().update_status();
        }

        fn signals() -> &'static [glib::subclass::Signal] {
            static SIGNALS: std::sync::OnceLock<Vec<glib::subclass::Signal>> =
                std::sync::OnceLock::new();

            SIGNALS.get_or_init(|| {
                vec![
                    // Emitted once per release found by a check the user did
                    // not ask for, so that the window can mention it. A check
                    // the user made themselves is answered by the row they
                    // pressed, and does not need announcing twice.
                    glib::subclass::Signal::builder("update-announced")
                        .param_types([String::static_type()])
                        .build(),
                ]
            })
        }
    }
}

glib::wrapper! {
    /// The state of this installation with respect to the release feed.
    pub struct Updates(ObjectSubclass<imp::Updates>);
}

impl Updates {
    /// Create a new `Updates`.
    pub(crate) fn new() -> Self {
        glib::Object::new()
    }

    /// Start the timer that makes automatic checks.
    ///
    /// Does nothing when this installation cannot update itself, so a Flatpak
    /// never contacts the feed at all.
    pub(crate) fn start(&self) {
        if !self.supported() {
            debug!("This installation does not update itself; not checking");
            return;
        }

        glib::timeout_add_local_once(
            FIRST_CHECK_DELAY,
            clone!(
                #[weak(rename_to = obj)]
                self,
                move || {
                    obj.check_if_due();

                    glib::timeout_add_local(
                        TIMER_INTERVAL,
                        clone!(
                            #[weak]
                            obj,
                            #[upgrade_or]
                            glib::ControlFlow::Break,
                            move || {
                                obj.check_if_due();
                                glib::ControlFlow::Continue
                            }
                        ),
                    );
                }
            ),
        );
    }

    /// Change the stored settings and write them back.
    fn edit_settings(&self, edit: impl FnOnce(&mut UpdateSettings)) {
        let mut settings = self.imp().settings.borrow_mut();

        edit(&mut settings);
        settings.save();
    }

    /// Check, but only if the settings say one is due.
    fn check_if_due(&self) {
        if self.imp().settings.borrow().is_check_due(now()) {
            self.check(false);
        }
    }

    /// Ask the feed what the current release is.
    ///
    /// `user_initiated` says whether somebody pressed a button: an automatic
    /// check announces what it finds, and says nothing at all when it fails,
    /// because a laptop that is offline is not a problem the user needs to be
    /// told about.
    pub(crate) fn check(&self, user_initiated: bool) {
        let imp = self.imp();

        if matches!(
            self.state(),
            UpdateState::Checking | UpdateState::Downloading | UpdateState::Installing
        ) {
            return;
        }

        if !self.supported() {
            return;
        }

        self.set_state(UpdateState::Checking);

        let channel = imp.settings.borrow().effective_channel();

        spawn!(clone!(
            #[weak(rename_to = obj)]
            self,
            async move {
                let result = spawn_tokio!(async move { updates::check(channel).await })
                    .await
                    .expect("the update check task is never aborted");

                obj.edit_settings(|settings| settings.last_check = Some(now()));

                match result {
                    Ok(check) => {
                        if let Some(release) = check.available {
                            let version = release.version.clone();
                            let announce = !user_initiated
                                && !obj.imp().settings.borrow().is_skipped(&release);

                            obj.imp().release.replace(Some(release));
                            obj.imp().available_version.replace(version.clone());
                            obj.notify_available_version();
                            obj.set_state(UpdateState::Available);

                            if announce {
                                obj.emit_by_name::<()>("update-announced", &[&version]);
                            }
                        } else {
                            obj.imp().release.replace(None);
                            obj.set_state(UpdateState::UpToDate);
                        }
                    }
                    Err(check_error) => {
                        // An automatic check that could not reach the feed is
                        // an ordinary thing on a laptop, so it is a log line
                        // rather than anything the user sees; one the user
                        // asked for owes them an answer.
                        if user_initiated {
                            error!("Could not check for updates: {check_error}");
                            obj.imp().status.replace(check_error.to_user_facing());
                            obj.set_state(UpdateState::Failed);
                            obj.notify_status();
                            return;
                        }

                        warn!("Could not check for updates: {check_error}");
                        obj.set_state(UpdateState::Idle);
                    }
                }
            }
        ));
    }

    /// Download the available release and start its installer.
    ///
    /// The application quits as the last step, because that is what both
    /// platforms need: Windows cannot replace an executable that is running,
    /// and macOS has already replaced the bundle by then and wants the new
    /// one started.
    pub(crate) fn install(&self) {
        let Some(release) = self.imp().release.borrow().clone() else {
            return;
        };

        let Some(asset) = release.asset_for_current_platform().cloned() else {
            return;
        };

        if matches!(
            self.state(),
            UpdateState::Downloading | UpdateState::Installing
        ) {
            return;
        }

        self.imp().progress.set(0.0);
        self.notify_progress();
        self.set_state(UpdateState::Downloading);

        let (sender, mut receiver) = mpsc::unbounded();

        // The download runs on a tokio worker and reports from there, which
        // is no place to touch a `GObject`. The numbers come back over a
        // channel and are applied here, on the main context.
        spawn!(clone!(
            #[weak(rename_to = obj)]
            self,
            async move {
                while let Some((downloaded, total)) = receiver.next().await {
                    let fraction = if total == 0 {
                        0.0
                    } else {
                        downloaded as f64 / total as f64
                    };

                    obj.imp().progress.set(fraction);
                    obj.notify_progress();
                    obj.update_status();
                }
            }
        ));

        spawn!(clone!(
            #[weak(rename_to = obj)]
            self,
            async move {
                let progress: Arc<dyn DownloadProgress> = Arc::new(ChannelProgress { sender });
                let dir = download_dir();

                let downloaded =
                    spawn_tokio!(
                        async move { updates::download(&asset, &dir, Some(progress)).await }
                    )
                    .await
                    .expect("the download task is never aborted");

                let path = match downloaded {
                    Ok(path) => path,
                    Err(download_error) => {
                        error!("Could not download the update: {download_error}");
                        obj.imp().status.replace(download_error.to_user_facing());
                        obj.set_state(UpdateState::Failed);
                        obj.notify_status();
                        return;
                    }
                };

                obj.set_state(UpdateState::Installing);

                if let Err(install_error) = start_installer(&path) {
                    error!("Could not start the installer: {install_error}");
                    obj.imp().status.replace(gettext(
                        "The update was downloaded but could not be installed.",
                    ));
                    obj.set_state(UpdateState::Failed);
                    obj.notify_status();
                    return;
                }

                // The installer is waiting for this process to be gone.
                crate::Application::default().quit();
            }
        ));
    }

    /// Where to read what the available release changed, if there is one.
    pub(crate) fn release_notes_url(&self) -> Option<String> {
        self.imp()
            .release
            .borrow()
            .as_ref()
            .map(|release| release.notes_url.clone())
    }

    /// Stop offering the release the last check found.
    pub(crate) fn skip(&self) {
        let Some(release) = self.imp().release.borrow().clone() else {
            return;
        };

        self.edit_settings(|settings| settings.skipped_version = Some(release.version));
    }

    /// Whether automatic checks are on.
    pub(crate) fn check_automatically(&self) -> bool {
        self.imp().settings.borrow().check_automatically
    }

    /// Turn automatic checks on or off.
    pub(crate) fn set_check_automatically(&self, enabled: bool) {
        if self.check_automatically() == enabled {
            return;
        }

        self.edit_settings(|settings| settings.check_automatically = enabled);
    }

    /// The channel being followed.
    pub(crate) fn channel(&self) -> Channel {
        self.imp().settings.borrow().effective_channel()
    }

    /// Follow the given channel, and look at it straight away.
    pub(crate) fn set_channel(&self, channel: Channel) {
        if self.channel() == channel {
            return;
        }

        self.edit_settings(|settings| {
            settings.channel = Some(channel);
            // A channel the user just chose is a channel they want an answer
            // about now, not tomorrow.
            settings.last_check = None;
            settings.skipped_version = None;
        });

        self.check(true);
    }

    /// Record a new state and rewrite the status sentence for it.
    fn set_state(&self, state: UpdateState) {
        if self.state() == state {
            return;
        }

        self.imp().state.set(state);
        self.update_status();
        self.notify_state();
    }

    /// Rewrite [`Self::status`] for the current state.
    fn update_status(&self) {
        let version = self.available_version();

        let status = match self.state() {
            UpdateState::Idle if !self.supported() => gettext(
                // Translators: shown where an update button would be, in a
                // build that is managed by something else.
                "Updates for this installation arrive through your app store.",
            ),
            UpdateState::Idle => gettext("Not checked yet."),
            UpdateState::Checking => gettext("Checking…"),
            UpdateState::UpToDate => gettext("Commune is up to date."),
            UpdateState::Available => gettext_f(
                // Translators: Do NOT translate the content between '{' and '}', this is a
                // variable name. {version} is a version number, like 1.0.
                "Commune {version} is available.",
                &[("version", &version)],
            ),
            UpdateState::Downloading => gettext_f(
                // Translators: Do NOT translate the content between '{' and '}', this is a
                // variable name. {percent} is a whole number, like 40.
                "Downloading Commune {version}… {percent}%",
                &[
                    ("version", &version),
                    ("percent", &format!("{:.0}", self.progress() * 100.0)),
                ],
            ),
            UpdateState::Installing => gettext("Installing. Commune will restart."),
            // Left as whatever the failing step put there.
            UpdateState::Failed => return,
        };

        self.imp().status.replace(status);
        self.notify_status();
    }
}

impl Default for Updates {
    fn default() -> Self {
        Self::new()
    }
}

/// A [`DownloadProgress`] that forwards the numbers to the main context.
struct ChannelProgress {
    /// Where the numbers go.
    sender: mpsc::UnboundedSender<(u64, u64)>,
}

impl DownloadProgress for ChannelProgress {
    fn progress(&self, downloaded: u64, total: u64) {
        // A closed channel means the receiving task is gone, which is not
        // worth interrupting a download for.
        let _ = self.sender.unbounded_send((downloaded, total));
    }
}

/// The current time as a Unix timestamp in seconds.
fn now() -> u64 {
    u64::try_from(glib::DateTime::now_utc().map_or(0, |now| now.to_unix())).unwrap_or_default()
}

/// Whether this installation can replace itself.
fn is_supported() -> bool {
    if !updates::is_supported() {
        return false;
    }

    #[cfg(target_os = "windows")]
    {
        crate::utils::windows_update::is_installed()
    }
    #[cfg(target_os = "macos")]
    {
        crate::utils::macos_update::is_installed()
    }
    #[cfg(not(any(target_os = "windows", target_os = "macos")))]
    {
        false
    }
}

/// Hand the downloaded artifact to the platform.
///
/// The name is checked first. The manifest is signed, so a `.txt` under the
/// `windows-x86_64-msi` key is a mistake rather than an attack — but it is a
/// mistake that would otherwise end with `msiexec` being pointed at
/// something that is not a package, and saying so plainly beats whatever the
/// installer would say about it.
#[allow(unused_variables, reason = "no platform to install on, on Linux")]
fn start_installer(path: &Path) -> std::io::Result<()> {
    #[cfg(target_os = "windows")]
    {
        if !crate::utils::windows_update::is_installer(path) {
            return Err(std::io::Error::other(
                "the release does not offer a Windows installer package",
            ));
        }

        crate::utils::windows_update::install(path)
    }
    #[cfg(target_os = "macos")]
    {
        if !crate::utils::macos_update::is_installer(path) {
            return Err(std::io::Error::other(
                "the release does not offer a macOS application archive",
            ));
        }

        crate::utils::macos_update::install(path)
    }
    #[cfg(not(any(target_os = "windows", target_os = "macos")))]
    {
        Err(std::io::Error::other(
            "this platform has no installer to hand the update to",
        ))
    }
}
